### DISCUSSION TITLE
Bug: Editor is altering code - Even when the source is non-editable

### DISCUSSION THREAD

**Discussion by Gregory1701** [codefori/vscode-ibmi#1645] [IBM i] [answered]
I've seen this several times, but have finally had the time to write up a simple program you can use to see the problem.

Here's the scenario:

1. I use Ctl+Alt+P to open a source member in our production library. You can see the problem in my little sample program, but the important thing here is that I'm opening a source member which I do not have authority to change, yet the editor has altered the source as it presented it to me. As I am sure you will agree, this simply cannot happen. I need to be able to trust that when I open a source member, I'm seeing it exactly as it exists in the last saved version.
2. But this bug can be demonstrated by simply saving this program, then reopening it and you will see the 16 has been moved all the way to the left.
3. The source member has an SQL statement where a number is one of the columns it selects.
4. The number will get shifted all the way to the left. While this doesn't hurt anything if the source is **FREE, if the source is not **FREE, this action has just caused a compile error. But again, the fundamental rule here is that the source member should not be altered.

```
**FREE

Dcl-ds resultInfo qualified;
  written_date Date;
  store char(10);
  store_number Uns(3);
End-ds;

Exec SQL
  Select date(),
         'Carle Place',
         16
    Into :resultInfo
    From sysibm.sysdummy1;

Return;
```

**Comment by worksofliam** [comment]
In your instance, what determines if the source is non-editable? Everything you open up in VS Code is editable, unless you explicitly tell your filter or connect to be readonly. Using Control+Alt+P opens things up as editable unless the connected is protected.

If you try and save a member that you do not have authority to, then it will fail, but you will still be able to edit it. We don't check individual member authority when opening with the shortcut.

**Reply by Gregory1701** [reply]
I may have complicated the issue by adding in the piece about me not having authority to the source member.

The root of the issue I'm trying to address is that by just opening a source, I don't want the source to be altered at all unless I alter it. So what's happening today is I can open the source, not make any changes myself and compare it to the original source and I already have differences. This cannot be. Additionally, in the event that the above test program was converted to a non-**FREE source member, then just opening the source will cause it to fail a compile.

**Comment by SJLennon** [comment]
FWIW:  On PUB400,COM I can't replicate the problem. 

I created the sample program in Code for IBM i and saved it.

In SEU browse it looks like this:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/5e77484d-39da-4dc4-adf8-7b4ef06781fc)

I opened it in Code for IBM i and it looks like this, with the 16 in the right place:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/f3673e77-97f0-4e66-9ace-f952f47b80be)

Even if I reformat the document to get rid of the linter errors, the 16 does not move:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/defdf743-5284-498a-89db-55639b9471b3)

(Is it possible there are some hidden characters in the line? Maybe dump the "16" line in hex?)

**Reply by Gregory1701** [reply]
I would hope there are not hidden characters as I wrote this program from scratch in VS Code. But looking at it in hex view, the line does look clean to me:
![image](https://github.com/codefori/vscode-ibmi/assets/132368707/1b0fbe29-c8cc-4318-bdae-edd35e0b7e1f)
But to reiterate my previous point, we can't have source changed when we open it. In other words, hidden characters or not, when I open it in SEU (gross!) it displays the source exactly as it is; no changes.

**Reply by SJLennon** [reply]
I still can't replicate the problem on PUB400 (the only machine I have access too.)

I created the program in Code for IBM i by cutting and pasting from you code above.  I have opened it from both the browser and with Ctrl-alt-p and see no problems.

Perhaps some other extension you have active may be interfering?   Maybe try disabling any extension related to SQL?

(Caveat, I'm not a developer of the extension, just a user.)

**Reply by Gregory1701** [reply]
Weird... I just disabled all extensions. Then re-enabled just Code for i. Same thing.

**Reply by worksofliam** [reply]
Due to the nature of this issue and size of this thread, step by step instructions to recreate this issue would be good.

**Reply by SJLennon** [reply]
Did you exit VS Code and restart? 

Have you looked at the OUTPUT tab Code for IBM i for anything odd? 
And in there is see things like this:

`system "CPYTOSTMF FROMMBR('/QSYS.lib/LENNONS1.lib/TEMP.file/GREGORY.mbr') TOSTMF('/tmp/lennons/code_for_IBM_i/vscodetemp-O_FSwk50bf') STMFOPT(*REPLACE) STMFCCSID(1208) DBFCCSID(*FILE)"`

You might want to look at your /tmp file and see if the 16 has moved there.

Any issues with CCSIDs?  

No other ideas at this point...

**Reply by Gregory1701** [reply]
This is my Output when I open the source. It's doing some kind of DB2 commands to open the source instead of  CPYTOSTRMF?

/home/PGMGREGORY: LC_ALL=EN_US.UTF-8 system "call QSYS/QZDFMDB2 PARM('-d' '-i' '-t')"
CREATE OR REPLACE ALIAS ILEDITOR.PGMGREGORY_QTESTSRC_SQLBUG for "PGMGREGORY"."QTESTSRC"("SQLBUG")
{
    "code": 0,
    "signal": null,
    "stdout": "DB2>\n  ?>\nDB20000I  THE SQL COMMAND COMPLETED SUCCESSFULLY.",
    "stderr": ""
}

/home/PGMGREGORY: LC_ALL=EN_US.UTF-8 system "call QSYS/QZDFMDB2 PARM('-d' '-i' '-t')"
SELECT LENGTH(srcdta) as LENGTH FROM ILEDITOR.PGMGREGORY_QTESTSRC_SQLBUG limit 1
{
    "code": 0,
    "signal": null,
    "stdout": "DB2>\n  ?>\n\nLENGTH     \n-----------\n        100\n\n  1 RECORD(S) SELECTED.",
    "stderr": ""
}

/home/PGMGREGORY: LC_ALL=EN_US.UTF-8 system "call QSYS/QZDFMDB2 PARM('-d' '-i' '-t')"
select srcdat, rtrim(translate(srcdta, x'2F', x'25')) as srcdta from ILEDITOR.PGMGREGORY_QTESTSRC_SQLBUG
{
    "code": 0,
    "signal": null,
    "stdout": "DB2>\n  ?>\n\nSRCDAT   SRCDTA                                                                                              \n-------- ----------------------------------------------------------------------------------------------------\n 231111  **FREE                                                                                              \n 231111                                                                                                      \n 231111  Dcl-ds resultInfo qualified;                                                                        \n 231111    written_date Date;                                                                                \n 231111    store char(10);                                                                                   \n 231111    store_number Uns(3);                                                                              \n 231111  End-ds;                                                                                             \n 231111                                                                                                      \n 231111  Exec SQL                                                                                            \n 231111    Select date(),                                                                                    \n 231111           'Carle Place',                                                                             \n 231111           16                                                                                         \n 231111      Into :resultInfo                                                                                \n 231111      From sysibm.sysdummy1;                                                                          \n 231111                                                                                                      \n 231111  Return;                                                                                             \n\n 16 RECORD(S) SELECTED.",
    "stderr": ""
}

/home/PGMGREGORY: LC_ALL=EN_US.UTF-8 system "call QSYS/QZDFMDB2 PARM('-d' '-i' '-t')"
CREATE OR REPLACE ALIAS ILEDITOR.PGMGREGORY_VSCODE_RPGLINT for "PGMGREGORY"."VSCODE"("RPGLINT")
{
    "code": 4,
    "signal": null,
    "stdout": "DB2>\n  ?>\n\n **** CLI ERROR *****\n         SQLSTATE: 42501\nNATIVE ERROR CODE: -551\nNot authorized to object PGMGR00002 in ILEDITOR type *FILE.",
    "stderr": ""
}

/home/PGMGREGORY: LC_ALL=EN_US.UTF-8 system "call QSYS/QZDFMDB2 PARM('-d' '-i' '-t')"
SELECT LENGTH(srcdta) as LENGTH FROM ILEDITOR.PGMGREGORY_VSCODE_RPGLINT limit 1
{
    "code": 4,
    "signal": null,
    "stdout": "DB2>\n  ?>\n\n **** CLI ERROR *****\n         SQLSTATE: 42704\nNATIVE ERROR CODE: -204\nVSCODE in PGMGREGORY type *FILE not found.",
    "stderr": ""
}

**Reply by Gregory1701** [reply, marked as answer]
This seems to have resolved itself. All it took was my old laptop having a corrupt kernel in the OS, getting a new laptop and fresh installing code4i. How many licks does it take to get to the center of a tootsie roll pop? The world may never know...
